### PR TITLE
STITCH-1958 - Implement atomic sync operations

### PR DESCRIPTION
**Pull request by edaniels**
General algorithm and implementation status:
For each operation
	1. Take out a collection level (NS) lock (done)
	2. For reads
		2a. do a normal read and release the lock (done)
	3. For writes
		3a. find documents to be affected (done)
		3b. Write the documents to an undo collection (done)
		3c. do the write to the local collection (done)
		3d. mark the document config as having an uncommitted write (done)
		3e. remove the undo document (done)
	4. Unlock the collection level lock (done)

For sync
	1. For each namespace to synchronize, take out a lock when wanting to making read/writes on doc configs/local collection (done)

On DataSynchronizer startup
	1. Lock the NS and dispatch a request to do the below (done)
	2. Search for any undo documents (done)
	3. Replace the local database version of the document with the undo document (done)
	4. Set difference of local docs and configs. Remove difference from local docs (done)
	5. Unlock the NS (done)

Drive-bys:
* insertOneAndSync -> insertOne
* insertManyAndSync -> insertMany
* Remove Server version of Sync
* Stop publishing Server version local MongoDB service

### CODE REVIEW

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 1]
```diff
@@ -2107,14 +2333,28 @@ private void deleteOneFromRemote(
...
+      if (config == null) {
+        return;
+      }
+      final BsonDocument documentToDelete = getLocalCollection(namespace)
+          .find(getDocumentIdFilter(documentId)).first();
+      if (documentToDelete == null) {
+        desyncDocumentFromRemote(namespace, documentId);
+        return;
+      }
+      getUndoCollection(namespace).insertOne(documentToDelete);
```
Comment (adamchel) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [nit] This applies throughout this file, but why not have a `final MongoCollection<..> undoColl = getUndoCollection(...)` or `final MongoCollection<...> localColl = getlocalColl(...)` in places you use the utility functions so you don't unnecessarily create the collection object from scratch every time you do an operation?

File: config/findbugs-exclude.xml [hunk 2]
```diff
@@ -9,4 +9,8 @@
     <Match>
         <Source name="~.*\.kt"/>
     </Match>
+    <Match>
+        <Class name="com.mongodb.stitch.core.services.mongodb.remote.sync.internal.DataSynchronizer" />
+        <Bug pattern="SC_START_IN_CTOR" />
```
[the remark this answers is not in the document]
Comment (edaniels) [reply config/findbugs-exclude.xml]: That worked!

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 3]
```diff
@@ -177,6 +169,70 @@ public DataSynchronizer(
...
+      undoCollection.deleteOne(filter);
+    }
+
+    // Find local documents for which there are no document configs and delete them. This covers
+    // inserts, upserts, and desync deletes.
+    localCollection.deleteMany(new BsonDocument(
+        "_id",
+        new BsonDocument(
+            "$nin",
+            new BsonArray(new ArrayList<>(getSynchronizedDocumentIds(nsConfig.getNamespace()))))));
```
[the remark this answers is not in the document]
Comment (edaniels) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: we haven't but we should for GA

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java (hunk 1 above)
Comment (edaniels) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: laziness. will address

File: config/findbugs-exclude.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (edaniels) [reply config/findbugs-exclude.xml]: nvm can't mock a final class and I don't want to add power mock

**Pull request merged**